File: comp_arch_assignment1/question2/ques_2_int.cpp
#include <iostream>
#include <vector>
#include <ctime>

// Function to multiply matrices
template <typename T>
void multiplyMatrices(const std::vector<std::vector<T> >& A, const std::vector<std::vector<T> >& B, std::vector<std::vector<T> >& C, int N) {
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            C[i][j] = 0;
            for (int k = 0; k < N; k++) {
                C[i][j] += A[i][k] * B[k][j];
            }
        }
    }
}

int main() {
    // Define the set of values for N
    int sizes[] = {64, 128, 256, 512, 1024};
    int num_sizes = sizeof(sizes) / sizeof(sizes[0]);

    for (int s = 0; s < num_sizes; s++) {
        int N = sizes[s];
        std::cout << "Matrix size: " << N << "x" << N << std::endl;

        // Define matrices
        std::vector<std::vector<int> > A(N, std::vector<int>(N, 1));
        std::vector<std::vector<int> > B(N, std::vector<int>(N, 2));
        std::vector<std::vector<int> > C(N, std::vector<int>(N, 0));

        struct timespec start, end;
        clock_gettime(CLOCK_MONOTONIC, &start);

        // Perform matrix multiplication
        multiplyMatrices(A, B, C, N);

        clock_gettime(CLOCK_MONOTONIC, &end);
        double time_taken = (end.tv_sec - start.tv_sec) + (end.tv_nsec - start.tv_nsec) / 1e9;
        std::cout << "Time taken by matrix multiplication: " << time_taken << " seconds" << std::endl;
        std::cout << std::endl;  // Add a newline for better readability
    }

    return 0;
}

File: comp_arch_assignment1/question2/ques_2_int.py
import time

# Function to multiply matrices without NumPy
def multiply_matrices(A, B, N):
    # Initialize the result matrix with zeros
    C = [[0 for _ in range(N)] for _ in range(N)]
    
    # Perform matrix multiplication
    for i in range(N):
        for j in range(N):
            for k in range(N):
                C[i][j] += A[i][k] * B[k][j]
    
    return C

# List of N values to iterate through
N_values = [64, 128, 256, 512, 1024]

for N in N_values:
    print(f"\nMatrix size: {N}x{N}")

    # Integer matrix multiplication
    A = [[1 for _ in range(N)] for _ in range(N)]
    B = [[2 for _ in range(N)] for _ in range(N)]

    # Measure wall-clock time and CPU time
    start_time = time.time()

    # Perform matrix multiplication
    C = multiply_matrices(A, B, N)

    end_time = time.time()
    

    print(f"Time taken by integer matrix multiplication: {end_time - start_time:.6f} seconds (Execution Time)")

    # Double matrix multiplication
    A = [[1.0 for _ in range(N)] for _ in range(N)]
    B = [[2.0 for _ in range(N)] for _ in range(N)]

    # Measure wall-clock time and CPU time
    start_time = time.time()
    

    # Perform matrix multiplication
    C = multiply_matrices(A, B, N)

    end_time = time.time()

    print(f"Time taken by double matrix multiplication: {end_time - start_time:.6f} seconds (Execution Time)")